engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java

```diff
@@ -895,8 +895,10 @@ private void redoPartitions(PMapping currentMapping, String note)
         addedPartitions.add(newPartition);
       } else {
         // check whether mapping was changed
+        int currentPartitionsSize = mainPC.currentPartitions.size();
         for (DefaultPartition<Operator> pi : mainPC.currentPartitions) {
-          if (pi == newPartition && pi.isModified()) {
+          if (pi == newPartition && (pi.isModified() ||
```

Comment:
@siyuanh A1->B1->C1 and A2->B2->C2 don't need to be redeployed assuming that they continue to handle the same set of keys. In your example, if S emits keys 1 and 2 and at some point starts to emit 1, 2 and 3 it is necessary to deploy A3->B3->C3 and redeploy U and Aggregator, while A1->B1->C1 and A2->B2->C2 don't need to be redeployed. In case, S starts to emit 1, 2, 3 and 4 and both 1 and 4 are routed to A1, A1->B1->C1 needs to be redeployed as it's key set changes.